mlir-sycl/lib/Conversion/SYCLToLLVM/SYCLToLLVM.cpp

```diff
@@ -354,6 +419,130 @@ class GetRangeSizePattern : public ConvertOpToLLVMPattern<Op> {
     rewriter.replaceOp(op, newValue);
   }
 };
+
+template <typename Op>
+class GetLinearIDPattern : public ConvertOpToLLVMPattern<Op> {
+protected:
+  using typename ConvertOpToLLVMPattern<Op>::OpAdaptor;
+  using ConvertOpToLLVMPattern<Op>::ConvertOpToLLVMPattern;
+
+  virtual Value getID(OpBuilder &builder, Location loc, Type ptrTy,
+                      Value thisArg, int32_t index) const = 0;
+  virtual Value getRange(OpBuilder &builder, Location loc, Type ptrTy,
+                         Value thisArg, int32_t index) const = 0;
+
+public:
+  LogicalResult match(Op) const override { return success(); }
+
+  void rewrite(Op op, OpAdaptor adaptor,
+               ConversionPatternRewriter &rewriter) const final {
+    const auto thisArg = adaptor.getOperands()[0];
+    const auto ptrTy = LLVM::LLVMPointerType::get(
+        op.getType(), thisArg.getType()
+                          .template cast<LLVM::LLVMPointerType>()
+                          .getAddressSpace());
+    const auto loc = op.getLoc();
+    const auto dimension = getDimensions(op.getOperand().getType());
+    const auto newValue = [this, loc, thisArg, ptrTy, dimension,
+                           &rewriter]() -> Value {
+      switch (dimension) {
+      case 1:
+        // get_id(0)
+        return getID(rewriter, loc, ptrTy, thisArg, 0);
+      case 2: {
+        // get_id(0) * get_range(1) + get_id(1)
+        const auto id0 = getID(rewriter, loc, ptrTy, thisArg, 0);
+        const auto r1 = getRange(rewriter, loc, ptrTy, thisArg, 1);
+        const Value prod = rewriter.create<arith::MulIOp>(loc, id0, r1);
+        const auto id1 = getID(rewriter, loc, ptrTy, thisArg, 1);
+        return rewriter.create<arith::AddIOp>(loc, prod, id1);
+      }
+      case 3: {
+        // get_id(0) * get_range(1) * get_range(2) + get_id(1) * get_range(2) +
+        // get_id(2)
+        const auto id0 = getID(rewriter, loc, ptrTy, thisArg, 0);
+        const auto r1 = getRange(rewriter, loc, ptrTy, thisArg, 1);
+        const Value prod0 = rewriter.create<arith::MulIOp>(loc, id0, r1);
+        const auto r2 = getRange(rewriter, loc, ptrTy, thisArg, 2);
+        const Value prod1 = rewriter.create<arith::MulIOp>(loc, prod0, r2);
+        const auto id1 = getID(rewriter, loc, ptrTy, thisArg, 1);
+        const Value prod2 = rewriter.create<arith::MulIOp>(loc, id1, r2);
+        const Value add = rewriter.create<arith::AddIOp>(loc, prod1, prod2);
+        const auto id2 = getID(rewriter, loc, ptrTy, thisArg, 2);
+        return rewriter.create<arith::AddIOp>(loc, add, id2);
+      }
+      default:
+        llvm_unreachable("Invalid number of dimensions");
+      }
+    }();
```

Comment:
Can you remove the lambda and instead just assign the newValue in the various switch cases ?